### PR TITLE
Allow async I/O in config files

### PR DESCRIPTION
**Pull request by kelunik**
This allows fetching e.g. configuration and building hosts based on it.

### CODE REVIEW

**Comment by kelunik** [comment]
@rdlowrey Maybe we should adjust the failure rate on Coveralls to `-0.25%`.

**Comment by benjamingr** [comment]
Or maybe write a test for it :D?

**Comment by kelunik** [comment]
Yes. :-D

**Comment by kelunik** [comment]
Maybe I should just enable the "require all checks passed before merge" option, leads to better PRs. :-)

File: lib/Bootstrapper.php [hunk 1]
```diff
@@ -16,16 +16,24 @@ public function __construct(callable $hostAggregator = null) {
...
      */
-    public function boot(Logger $logger, Console $console): Server {
+    public function boot(Logger $logger, Console $console): \Generator {
         $configFile = $this->selectConfigFile((string) $console->getArg("config"));
         $logger->info("Using config file found at $configFile");
-        if (!include($configFile)) {
+
+        $returnValue = include $configFile;
+
+        if (!$returnValue) {
```
Comment (staabm) [inline lib/Bootstrapper.php]: wouldnt it lead to a problem when the config doesnt return anything?

Comment (kelunik) [inline lib/Bootstrapper.php]: In that case `(int) 1` is returned for success and `false` for failure.

File: test/BootstrapperTest.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+use function Amp\resolve;
+use function Amp\wait;
+
+class BootstrapperTest extends \PHPUnit_Framework_TestCase {
+    /**
+     * @test
+     * @expectedException \DomainException
+     * @expectedExceptionMessage No config file found, specify one via the -c switch on command line
+     */
+    public function throwsWithoutConfig() {
```
Comment (bwoebi) [inline test/BootstrapperTest.php]: This test has a problem: it may spuriously fail in case you have a config in the default locations.

Comment (kelunik) [inline test/BootstrapperTest.php]: I know, but I'd say we should disable default locations anyway.

**Comment by bwoebi** [comment]
In general, PHP doesn't allow you to yield inside a file. So, when we want async I/O in config file, we need to do a big `return (function() { /* .. */ })();`, right?

File: test/BootstrapperTest.php (hunk 2 above)
Comment (bwoebi) [inline test/BootstrapperTest.php]: I don't disagree… let me do this and merge this PR.

**Comment by kelunik** [comment]
You missed the coroutine wrapper, but right.

**Pull request merged**